EIPS/eip-5994.md

```diff
@@ -0,0 +1,85 @@
+
+## Motivation
+
+This Token Pods concept allows to have permissionless extensions for tokens, which can be used for incentivisations (aka farms) or delegations or any other ideas.
+
+Example implementation can be found here:
+- https://github.com/1inch/erc20-pods
+
+Examples of Pods:
+- https://github.com/1inch/farming
+- https://github.com/1inch/delegating
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+```solidity
+interface IPod {
+    function updateBalances(address from, address to, uint256 amount) external;
+}
+
+interface IERC20Pods is IERC20 {
+    function hasPod(address account, address pod) external view returns(bool);
+    function podsCount(address account) external view returns(uint256);
+    function podAt(address account, uint256 index) external view returns(address);
+    function pods(address account) external view returns(address[] memory);
+    function podBalanceOf(address pod, address account) external view returns(uint256);
+
+    function addPod(address pod) external;
+    function removePod(address pod) external;
+    function removeAllPods() external;
+}
+
+interface IERC721Pods is IERC721 {
+    function hasPod(address account, address pod) external view returns(bool);
+    function podsCount(address account) external view returns(uint256);
+    function podAt(address account, uint256 index) external view returns(address);
+    function pods(address account) external view returns(address[] memory);
+    function podBalanceOf(address pod, address account) external view returns(uint256);
+
+    function addPod(address pod) external;
+    function removePod(address pod) external;
+    function removeAllPods() external;
+}
+```
+
+### Pod specification
+
+- Pod MUST implement interface `IPod` and fit execution of `updateBalances` method in 200,000 gas and never revert.
+
+- Pod SHOULD make sure `updateBalances(...)` method was triggered by `token` it was designated for.
+
+- Pod can track changes of balances of users who voluntarily added this pod, all other account will be represented as `address(0)` and interactions with such accounts could be considered as mints and burns.
+
+- Pod SHOULD NOT try to change balances of the token it was designated for, since it would introduce reentrancy, which will not pass.
+
+## Rationale
+
+## Backwards Compatibility
+
+Fully backward compatibile with ERC20 and ERC721.
+
+## Security Considerations
+
+Pods implementations will never block balance update, but can eat upto 200,000 gas. Number of added Pods per account is limited to some immutable value (reasonable amount taking into account max transaction gas limit).
```

Comment:
* ERCs MUST NOT depend on gas costs. They change from one hard fork to the next. EIP-165 is the one exception to this as it was finalized before that rule was put in place. It's well-recognized that the 30,000 gas limit is not a "real" rule, however.
* This isn't in the spec!

Reply:
I see no way to avoid this constant. This number should be specified. Moreover we have restriction of number of pods added to one account. N*200k should be less than max transaction gas limit. So I expect devs to use 200k constant and 10-20 pods limit per account.

Reply:
Core concept is permissionless pods and risk-free participation. So Pods cant block transfers/mint/burns of designated assets.

Reply:
It's okay to have the gas limit as a should. But it is *not* okay to have it as a must.